grep-0011-scheduler.md

```diff
@@ -0,0 +1,124 @@
+data transfers are usually not taken into account. But their behavior can be 
+detrimental in the performance of the KPN due to cache misses, interrupts, delays, 
+etc. 
+
+Some background on traditional scheduling can be found in [6] [7] [8] [9]
+
+### The role of the Criticality function
+Contrary to priority schedulers, the Criticality function takes the differences 
+in execution speed in heterogeneous environments into account. It allows the 
+scheduler to analyze which KPN node must be scheduled for optimal performance.
+
+I stress the use of heterogeneous processing environments because the 
+homogeneous environment is a simplified version of it. Also, many GR users do have 
+options to execute the KPN on other processors next to their main CPU. Furthermore
+it helps software engineers focus their effort in optimisations (using SIMD, 
+other languages or better compilers). 
+
+No prior implementation of a Criticality-type scheduler, a scheduler using the 
+Criticality function, exists. So it has to be made from scratch. All tooling, 
+needed to analyze the performance of the Gnuradio application blocks need to be 
+created too. This is necessary to calculate the Criticality function for the
+scheduler. 
+
+After creating the Criticality function for a KPN, other (sometimes conflicting) 
+goals exist, such as having the scheduler:
+- Optimize for low memory usage
+- Optimize for cache hit rate at processors
+- Optimize for latency
+- Optimize for data throughput (at various data busses)
+- Optimize for execution time jitter
+
+If a user wants to accelerate the Gnuradio application, optimalisation needs to
+be done. It would be useful if the scheduler can analyze the following for a user:
+- What processes can be optimized and explain the required speedup needed to
+  attain maximum performance. 
+- Predict performance (bottlenecks) on multiple architectures
+
+At some point, the GR scheduler must cooperate with the scheduler of the operating
+system. Some consideration must also be taken into account for the differences 
+of optimization that can be done. E.g. a RT kernel has different options for
+high performance than a normal kernel. Other workloads of a system come into play
+as well (e.g. a browser from a user). 
+  
+## Description
+
+### Origins
+There are a number of papers on GRCon17 describing that enhancements in scheduling
+can be worthwile [2][3][4][5]. Some information on this topic was shared at FOSDEM18
+with the remark of Marcus Müller that scheduling is insufficient at the moment.
+Coincidentally, I was there and I wrote a thesis on 'Exploring criticality
+numbers for Kahn Process Networks' and proposed a Criticality function as described
+in [1]. This describes how Kahn Process Networks can be analyzed to show what block
+is critical and what block is not (and by what factor). 
+
+
+### Current status
+Currently, no scheduler exists that uses the Criticality function. 
+
+Todo:
```

Comment:
Clearly, we need to expand this section :grin:

Reply:
Hi Martin,

I processed the typo's and title. I agree on that.

So to address some of your questions:

Success criterion: I would say that total execution time is most often used for a given "task (set)". But it would need to be repeatable which is pretty hard given the fact that most people have different computer setups. One of the things you need is some sort of profiler and some sort of benchmark. To be able to compare that. 

Normal problems with profilers is that you need a lot of time to log the results. It needs to be done in such a way that this does not interfere too much with results. 

Secondary criteria are: 
- Maximum and average memory usage
- Maximum and average delay (from inputs to outputs)

It may be that some application would like these criteria swapped. Or requires others. But lets stick to this first. All that should influence scheduling

Pluggable: Yes please. The good thing is that scheduling is usually broken down into "the algorithm" and the part that does the stuff, the "dispatcher". 

Action plan:
- Create profiler. This probably requires the scheduler to be pluggable already. 
- Profile the applications and the FIFO's. 
- I think I can model the FIFO's too and create a scheduling optimisation based on that. 
- Create new scheduler. For now my idea is that the scheduler is static. 

Yours,


David